Fix in the MPA stub building code

sviret opened the pull request:
#### PR description:

This PR implements a fix in the stub building code of the L1Trigger/TrackTrigger package. Only one file is affected (TTStubBuilder.cc), and one line in this file. The PS modules can build up to 2 stubs per seed. It was set to one until now. So this is a very minor change (1->2)

#### PR validation:

The code change was tested with a private CMSSW version. As expected a small increase of the stub rate was observed in the PS module (less than 5% in any cases @ PU200).

Commit:
Fix in the MPA stub building code

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-34204/23429

- This PR adds an extra 16KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-34204/23429/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-34204/23429/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

sviret committed:
Update TTStubBuilder.cc

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-34204/23431

- This PR adds an extra 16KB to repository

cmsbuild commented:
A new Pull Request was created by  @sviret (Seb Viret) for master.

It involves the following packages:

L1Trigger/TrackTrigger


@cmsbuild, @rekovic, @srimanob, @cecilecaillol, @kpedro88 can you please review it and eventually sign? Thanks.
@Martin-Grunewald, @erikbutz, @skinnari this is something you requested to watch as well.
@silviodonato, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

srimanob commented:
Please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-198353/16160/summary.html
**COMMIT**: 6438d09d124957e4ebaed527653e9782ba98fef7
**CMSSW**: CMSSW_12_0_X_2021-06-22-1100/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/34204/16160/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 38
 - DQMHistoTests: Total histograms compared: 2785631
 - DQMHistoTests: Total failures: 357
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2785252
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 37 files compared)
 - Checked 160 log files, 37 edm output root files, 38 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_0_X_2021-06-22-1100+198353/43600/triggerResults): no differences found

srimanob commented:
Hi @sviret 

Could you please confirm if all changes see in the Phase-2 workflow (SiOuterTracker/Tracks, /Stubs) are expected?
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_0_X_2021-06-22-1100+198353/43600/validateJR.html

I assume yes, but better to x-check with you first. Thanks.

sviret commented:
Hi @srimanob,

Those changes are indeed expected. The stub rate in the PS modules is increased by few %, this is what is detected by the DQM.

cecilecaillol commented:
+l1

srimanob commented:
+Upgrade

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @silviodonato, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

qliphy commented on L1Trigger/TrackTrigger/plugins/TTStubBuilder.cc:
```diff
@@ -210,7 +210,7 @@ void TTStubBuilder<Ref_Phase2TrackerDigi_>::produce(edm::Event& iEvent, const ed
 
       /// Here tempOutput stores all the stubs from this lower cluster
       /// Check if there is need to store only one (if only one already, skip this step)
```
Should we also update the commented line accordingly?

sviret replied:
@qliphy thanks for the comment which actually pointed me toward a small bug in the code. I will repush a version with comments and the fix (which is just one line the loop). So please wait for the update before pushing the PR

sviret replied:
I'm currently running some quick checks but the impact should be negligible. The case where there is more than 2 stubs formed from one seed in a PS module was not accounted correctly (only 1 stub was kept). This should be very rare anyway, but better to implement the exact behavior

sviret committed:
Update TTStubBuilder.cc

Added some comments and correct a minor bug (impact on total stub PS rates negligible (test made))

sviret commented:
I made the correction and added some comments, as suggested. As expected impact on total stub rates is negligible, but now the code represents the correct behavior

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-34204/23474

- This PR adds an extra 16KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-34204/23474/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-34204/23474/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

sviret committed:
Update TTStubBuilder.cc

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-34204/23475

- This PR adds an extra 16KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-34204/23475/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-34204/23475/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

sviret committed:
Update TTStubBuilder.cc

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-34204/23476

- This PR adds an extra 20KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-34204/23476/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-34204/23476/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

sviret committed:
Update TTStubBuilder.cc

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-34204/23478

- This PR adds an extra 20KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-34204/23478/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-34204/23478/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

sviret committed:
Update TTStubBuilder.cc

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-34204/23479

- This PR adds an extra 20KB to repository

cmsbuild commented:
Pull request #34204 was updated. @cmsbuild, @rekovic, @srimanob, @cecilecaillol, @kpedro88 can you please check and sign again.

srimanob commented:
Please test

cmsbuild commented:
-1

**Failed Tests**: RelVals-INPUT 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-198353/16200/summary.html
**COMMIT**: 5d05e8c193dc5492d1591414b533122c0ee00b76
**CMSSW**: CMSSW_12_0_X_2021-06-23-1100/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/34204/16200/install.sh` to create a dev area with all the needed externals and cmssw changes.


## RelVals-INPUT



## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 38
 - DQMHistoTests: Total histograms compared: 2785631
 - DQMHistoTests: Total failures: 351
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2785258
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 37 files compared)
 - Checked 160 log files, 37 edm output root files, 38 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_0_X_2021-06-23-1100+198353/43632/triggerResults): no differences found

sviret commented:
Error seems to be the following:

2021/06/23 17:45:05 ERROR failed to parse X509 proxy: crypto/tls: failed to parse key: asn1: syntax error: sequence truncated

TMO this has nothing to do with the PR
@srimanob I suspect the solution is to just relaunch the test, no?

qliphy commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-198353/16217/summary.html
**COMMIT**: 5d05e8c193dc5492d1591414b533122c0ee00b76
**CMSSW**: CMSSW_12_0_X_2021-06-23-1600/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/34204/16217/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



The workflows 140.53 have different files in step1_dasquery.log than the ones found in the baseline. You may want to check and retrigger the tests if necessary. You can check it in the "files" directory in the results of the comparisons

**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 1272 differences found in the comparisons
 - DQMHistoTests: Total files compared: 38
 - DQMHistoTests: Total histograms compared: 2785631
 - DQMHistoTests: Total failures: 4030
 - DQMHistoTests: Total nulls: 20
 - DQMHistoTests: Total successes: 2781559
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -45.707 KiB( 37 files compared)
 - DQMHistoSizes: changed ( 140.53 ): -44.531 KiB      Hcal/DigiRunHarvesting
 - DQMHistoSizes: changed ( 140.53 ): -1.172 KiB       RPC/DCSInfo
 - DQMHistoSizes: changed ( 312.0 ): -0.004 KiB       MessageLogger/Warnings
 - Checked 160 log files, 37 edm output root files, 38 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_0_X_2021-06-23-1600+198353/43646/triggerResults): no differences found

cecilecaillol commented:
+l1

srimanob commented:
+Upgrade

- Change in Phase-2 wfs as expected.
- The total failure increases due to the issue from the input files used in workflow 140.53 (2011 HI).

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @silviodonato, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

qliphy commented:
+1
